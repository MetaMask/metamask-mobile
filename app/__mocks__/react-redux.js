export * from 'react-redux-test';
