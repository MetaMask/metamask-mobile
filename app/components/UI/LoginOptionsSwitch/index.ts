export { default as LoginOptionsSwitch } from './LoginOptionsSwitch';
