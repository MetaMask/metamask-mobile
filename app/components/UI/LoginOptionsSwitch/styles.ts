/* eslint-disable import/prefer-default-export */
import { fontStyles } from '../../../styles/common';
import { StyleSheet } from 'react-native';

export const createStyles = (colors: any) =>
  StyleSheet.create({
    screen: {
      flex: 1,
      backgroundColor: colors.background.default,
    },
    container: {
      position: 'relative',
      marginTop: 20,
      marginBottom: 30,
    },
    label: {
      flex: 1,
      fontSize: 16,
      color: colors.text.default,
      ...fontStyles.normal,
    },
    switch: {
      position: 'absolute',
      top: 0,
      right: 0,
    },
  });
