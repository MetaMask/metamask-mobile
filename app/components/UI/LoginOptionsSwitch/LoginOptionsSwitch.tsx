import React, { useCallback, useState } from 'react';
import { View, Switch, Text } from 'react-native';
import { mockTheme, useAppThemeFromContext } from '../../../util/theme';
import { strings } from '../../../../locales/i18n';
import { BIOMETRY_TYPE } from 'react-native-keychain';
import { createStyles } from './styles';
import {
  LOGIN_WITH_BIOMETRICS_SWITCH,
  LOGIN_WITH_REMEMBER_ME_SWITCH,
} from '../../../constants/test-ids';
import { useSelector } from 'react-redux';

interface Props {
  shouldRenderBiometricOption: BIOMETRY_TYPE | null;
  biometryChoiceState: boolean;
  onUpdateBiometryChoice: (biometryEnabled: boolean) => void;
  onUpdateRememberMe: (rememberMeEnabled: boolean) => void;
}

/**
 * View that renders the toggle for login options
 * The highest priority login option is biometrics and will always get rendered over other options IF it is enabled.
 * If the user has enabled login with remember me in settings and has turned off biometrics then remember me will be the option
 * If both of these features are disabled then no options will be rendered
 */
const LoginOptionsSwitch = ({
  shouldRenderBiometricOption,
  biometryChoiceState,
  onUpdateBiometryChoice,
  onUpdateRememberMe,
}: Props) => {
  const { colors } = useAppThemeFromContext() || mockTheme;
  const styles = createStyles(colors);
  const allowLoginWithRememberMe = useSelector(
    (state: any) => state.security.allowLoginWithRememberMe,
  );
  const [rememberMeEnabled, setRememberMeEnabled] = useState<boolean>(false);
  const onBiometryValueChanged = useCallback(
    async (newBiometryChoice: boolean) => {
      onUpdateBiometryChoice(newBiometryChoice);
    },
    [onUpdateBiometryChoice],
  );

  const onRememberMeValueChanged = useCallback(async () => {
    onUpdateRememberMe(!rememberMeEnabled);
    setRememberMeEnabled(!rememberMeEnabled);
  }, [onUpdateRememberMe, rememberMeEnabled]);

  // should only render remember me option if biometrics are disabled and rememberOptionMeEnabled is enabled in security settings
  // if both are disabled then this component returns null
  if (shouldRenderBiometricOption !== null) {
    return (
      <View style={styles.container}>
        <Text style={styles.label}>
          {strings(
            `biometrics.enable_${shouldRenderBiometricOption.toLowerCase()}`,
          )}
        </Text>
        <Switch
          onValueChange={onBiometryValueChanged}
          value={biometryChoiceState}
          style={styles.switch}
          trackColor={{
            true: colors.primary.default,
            false: colors.border.muted,
          }}
          thumbColor={colors.white}
          ios_backgroundColor={colors.border.muted}
          testID={LOGIN_WITH_BIOMETRICS_SWITCH}
        />
      </View>
    );
  } else if (shouldRenderBiometricOption === null && allowLoginWithRememberMe) {
    return (
      <View style={styles.container}>
        <Text style={styles.label}>
          {strings(`choose_password.remember_me`)}
        </Text>
        <Switch
          onValueChange={onRememberMeValueChanged}
          value={rememberMeEnabled}
          style={styles.switch}
          trackColor={{
            true: colors.primary.default,
            false: colors.border.muted,
          }}
          thumbColor={colors.white}
          ios_backgroundColor={colors.border.muted}
          testID={LOGIN_WITH_REMEMBER_ME_SWITCH}
        />
      </View>
    );
  }
  return null;
};

export default React.memo(LoginOptionsSwitch);
