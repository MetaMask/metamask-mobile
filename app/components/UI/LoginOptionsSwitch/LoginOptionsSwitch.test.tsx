import React from 'react';
import { shallow } from 'enzyme';
import LoginOptionsSwitch from './LoginOptionsSwitch';
import { BIOMETRY_TYPE } from 'react-native-keychain';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
describe('LoginWithBiometricsSwitch', () => {
  const mockStore = configureMockStore();
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const handleUpdate = (_biometricsEnabled: boolean) => {};
  it('should render correctly', () => {
    const store = mockStore({});
    const wrapper = shallow(
      <Provider store={store}>
        <LoginOptionsSwitch
          shouldRenderBiometricOption={BIOMETRY_TYPE.FACE}
          biometryChoiceState
          onUpdateBiometryChoice={handleUpdate}
          onUpdateRememberMe={handleUpdate}
        />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should return empty object when shouldRenderBiometricOption is undefined and allowLoginWithRememberMe is false in settings', () => {
    const store = mockStore({});
    const wrapper = shallow(
      <Provider store={store}>
        <LoginOptionsSwitch
          onUpdateBiometryChoice={handleUpdate}
          onUpdateRememberMe={handleUpdate}
          biometryChoiceState
        />
      </Provider>,
    );
    expect(wrapper).toMatchObject({});
  });
});
