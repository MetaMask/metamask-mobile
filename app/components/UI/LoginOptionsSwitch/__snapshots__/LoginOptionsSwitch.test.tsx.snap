// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginWithBiometricsSwitch should render correctly 1`] = `
<Component
  biometryChoiceState={true}
  onUpdateBiometryChoice={[Function]}
  onUpdateRememberMe={[Function]}
  shouldRenderBiometricOption="Face"
/>
`;
